### PR TITLE
Param tests

### PR DESCRIPTION
**Pull request by MTCam**
* adds a simple option to write the code to file for human inspection
* parametrize the tests so that they hit all the available mechanisms

@inducer, @ecisneros8

### CODE REVIEW

File: pyrometheus/__init__.py [hunk 1]
```diff
@@ -555,7 +555,10 @@ def gen_python_code(sol: ct.Solution):
...
                                             for r in sol.reactions()])),
     )
-    print(code)
+    if file is not None:
+        print(code, file=file)
+    else:
+        print(code)
     exec_dict = {}
     exec(compile(code, "<generated code>", "exec"), exec_dict)
     return exec_dict["Thermochemistry"]
```
Comment (inducer) [inline pyrometheus/__init__.py:564]: This is just factored wrong. There should be a function that returns the generated string.

File: pyrometheus/__init__.py [hunk 2]
```diff
@@ -555,7 +555,10 @@ def gen_python_code(sol: ct.Solution):
                                            [isinstance(r, ct.ThreeBodyReaction)
                                             for r in sol.reactions()])),
     )
-    print(code)
+    if file is not None:
+        print(code, file=file)
+    else:
+        print(code)
     exec_dict = {}
     exec(compile(code, "<generated code>", "exec"), exec_dict)
```
Comment (inducer) [inline pyrometheus/__init__.py:563]:
```suggestion
    exec(compile(code, "<generated code>", "exec"), exec_dict)
    exec_dict["_MODULE_SOURCE_CODE"] = code
```
Using this, you can see the generated code in https://documen.tician.de/pudb/.

Comment (MTCam) [reply pyrometheus/__init__.py:563]: This was more about getting the actual py file than it was about putting eyes on the code.  It isn't essential.

**Comment by MTCam** [comment]
I think this is what you had in mind @inducer?

**Comment by inducer** [comment]
7e11e6a

**Pull request merged**